hpcviewer: new package

mwkrentel opened the pull request:
Add binary package for hpcviewer and hpctraceviewer for the Rice
hpctoolkit on Linux x86_64, ppc64 and ppc64le.

----------

I may be pushing the boundaries of flake with the long sha256 strings.
But folding them to the next line is ugly.  So, we'll see how the
judge rules.


Also, hpcviewer depends on java, but the tcl module that spack
created for hpcviewer doesn't depend on java.  I was expecting that
`module load hpcviewer` would also load jdk as a dependency, but it
didn't.

The module file sets `PATH` and `CMAKE_PREFIX_PATH` and prints a help
string.  Nothing else, no mention of java.

Did I miss something, or do tcl modules just not support dependencies?

adamjstewart commented:
> I may be pushing the boundaries of flake with the long sha256 strings.

Spack automatically ignores checksums in flake8: https://github.com/spack/spack/blob/develop/lib/spack/spack/cmd/flake8.py#L81

> I was expecting that `module load hpcviewer` would also load jdk as a dependency, but it didn't.

See #8639. You can configure Spack to load run deps for certain packages, but it currently doesn't do this by default.

adamjstewart commented on var/spack/repos/builtin/packages/hpcviewer/package.py:
```diff
@@ -0,0 +1,69 @@
+    # Both hpcviewer and trace viewer have an install script.
+    def install(self, spec, prefix):
+        args = [
+            '--java', spec['java'].prefix,
```
You should use `spec['java'].home`. See the `jdk` package for an explanation why.

mwkrentel commented:
@adamjstewart Thanks!  I'm not yet done absorbing all the prior
discussion.  But I have one other question.

Flake complained about how I wrote the `viewer_url` and `trace_url`
support functions.

```
var/spack/repos/builtin/packages/hpcviewer/package.py:24:
[N805] first argument of a method should be named 'self'

var/spack/repos/builtin/packages/hpcviewer/package.py:30:
[N805] first argument of a method should be named 'self'
```

Is this specific to spack or is this python in general, that is,
defining a function inside a class?

You can see how I use these, right?  I can't really add a 'self'
argument because I don't think self is available at the point where I
call the function.  (Unless it's just a dummy argument that it
ignores.)

I could fold the format string into the version command, but that
seemed cluttered.  I thought it was cleaner to keep them separate.

If it was only the viewer, I could use `url_for_version`.  But I still
need something for the hpctraceviewer resource and I don't think there
exists a `url_for_resource` method.

Can you suggest something that flake would accept and still preserve
the "separation of concerns" that I've written?

adamjstewart commented:
> Is this specific to spack or is this python in general, that is, defining a function inside a class?

This is a Python thing

> I can't really add a 'self' argument because I don't think self is available at the point where I call the function. (Unless it's just a dummy argument that it ignores.)

`self` is a reference to the object itself. All class methods should have self as their first argument.

You can either add `self`, or if you don't think you need it, move the functions outside of the class.

mwkrentel commented:
@adamjstewart  Ok, I changed `spec['java'].prefix` to `.home` and updated
the ibm-java package to support that.

And I moved the viewer_url and trace_url functions outside of the
class, and now flake is happy (and happy with the right number of
blank lines).

----------

As for downloads that depend on machine type, yes, what I need in this
case is access to the arch (or target) type and use it to choose the
URL and checksum.

I guess this must break some fundamental spack design to provide full
access to the arch value inside every package?  That's why you
suggested `version('...', when='...')` ?

The difference, of course, is that python platform.machine() is the
build machine, whereas spack arch would allow specifying the target
machine ('host' in autoconf-speak), for example, back-end compute
nodes.

From a practical view, this isn't a problem for hpcviewer because you
only ever run the viewer on the front end.

But from a theoretical view, this is a major hole in spack.

But I'm content to leave that for another day.

adamjstewart commented on var/spack/repos/builtin/packages/hpcviewer/package.py:
```diff
@@ -0,0 +1,70 @@
+        version(ver, url=viewer_url(ver, mach), sha256=viewer_sha[(ver, mach)])
+
+        resource(name='hpctraceviewer', url=trace_url(ver, mach),
+                 sha256=trace_sha[(ver, mach)], destination='TRACE')
```
What happens when you run this on an arch that isn't supported? You would get a KeyError, right? Might want to check if the arch is a key in the dict before using it.

mwkrentel commented:
It will certainly fail with some kind of error.  The thematic spack
way of handling this is with conflicts.  But the right conflicts line
is something you currently can't write.

```
conflicts('not (target=x86_64 or target=ppc64)')
```

But I could add a test for machine type and at least fail with a more
informative error message.

adamjstewart commented:
The problem with conflicts is that Spack will literally crash just trying to read the file. Running `spack list` or any other command that accesses the files will cause it to read the file description.

mwkrentel commented:
I see now that I've seriously misunderstood the impact of the failure.
Even `spack info` would crash on an unsupported machine, which is
certainly not what I intended.  (It's all this implicit timing on when
things get run.  Sorry, my bad.)

I'll put something together tomorrow.  It needs to always be able to
create an Hpcviewer object on any arch so that `spack info`, etc
always work.  But `spack install` and even `spack spec` should fail
with a useful message on unsupported machines.

Thanks!  Disaster averted.

mwkrentel commented:
I fixed the dictionary lookup.  On unsupported machines, it just
creates no versions instead of crashing.  Spack seems to be ok with a
package with no versions, at least I didn't find anything that failed.

Again, sorry for my blind spot.

Anything else?

adamjstewart commented:
Nope, I think this is good to go.

mwkrentel commented:
ping @adamjstewart  Travis has finished, all green.
Thanks again!
